File: services/frontend/src/types/measurement.ts
/**
 * Body measurement type definitions matching the FastAPI backend models.
 */

export interface BodyMeasurement {
  id: number;
  date: string;
  weight_kg: number | null;
  body_fat_pct: number | null;
  chest_cm: number | null;
  waist_cm: number | null;
  hips_cm: number | null;
  bicep_left_cm: number | null;
  bicep_right_cm: number | null;
  thigh_left_cm: number | null;
  thigh_right_cm: number | null;
  neck_cm: number | null;
  shoulders_cm: number | null;
  forearm_cm: number | null;
  calf_cm: number | null;
  notes: string | null;
  created_at: string;
}

export interface CreateBodyMeasurement {
  date: string;
  weight_kg?: number | null;
  body_fat_pct?: number | null;
  chest_cm?: number | null;
  waist_cm?: number | null;
  hips_cm?: number | null;
  bicep_left_cm?: number | null;
  bicep_right_cm?: number | null;
  thigh_left_cm?: number | null;
  thigh_right_cm?: number | null;
  neck_cm?: number | null;
  shoulders_cm?: number | null;
  forearm_cm?: number | null;
  calf_cm?: number | null;
  notes?: string | null;
}

export interface MeasurementProgressPoint {
  date: string;
  value: number;
}

export interface MeasurementProgress {
  metric: string;
  unit: string;
  data: MeasurementProgressPoint[];
}

export type MeasurementMetric =
  | 'weight'
  | 'body_fat'
  | 'chest'
  | 'waist'
  | 'hips'
  | 'bicep_left'
  | 'bicep_right'
  | 'thigh_left'
  | 'thigh_right'
  | 'neck'
  | 'shoulders'
  | 'forearm'
  | 'calf';

export const MEASUREMENT_METRICS: { value: MeasurementMetric; label: string; field: keyof BodyMeasurement }[] = [
  { value: 'weight', label: 'Weight', field: 'weight_kg' },
  { value: 'body_fat', label: 'Body Fat %', field: 'body_fat_pct' },
  { value: 'chest', label: 'Chest', field: 'chest_cm' },
  { value: 'waist', label: 'Waist', field: 'waist_cm' },
  { value: 'hips', label: 'Hips', field: 'hips_cm' },
  { value: 'shoulders', label: 'Shoulders', field: 'shoulders_cm' },
  { value: 'neck', label: 'Neck', field: 'neck_cm' },
  { value: 'bicep_left', label: 'Bicep (L)', field: 'bicep_left_cm' },
  { value: 'bicep_right', label: 'Bicep (R)', field: 'bicep_right_cm' },
  { value: 'forearm', label: 'Forearm', field: 'forearm_cm' },
  { value: 'thigh_left', label: 'Thigh (L)', field: 'thigh_left_cm' },
  { value: 'thigh_right', label: 'Thigh (R)', field: 'thigh_right_cm' },
  { value: 'calf', label: 'Calf', field: 'calf_cm' },
];

File: services/shared/models/measurement.py
"""Body measurement models for tracking weight, body fat %, and tape measurements.

Used for logging body measurements over time and viewing progress charts.
"""

from datetime import date, datetime

from pydantic import BaseModel, Field


class BodyMeasurementCreate(BaseModel):
    """Request body for logging a body measurement entry."""

    date: date
    weight_kg: float | None = Field(default=None, ge=0)
    body_fat_pct: float | None = Field(default=None, ge=0, le=100)
    chest_cm: float | None = Field(default=None, ge=0)
    waist_cm: float | None = Field(default=None, ge=0)
    hips_cm: float | None = Field(default=None, ge=0)
    bicep_left_cm: float | None = Field(default=None, ge=0)
    bicep_right_cm: float | None = Field(default=None, ge=0)
    thigh_left_cm: float | None = Field(default=None, ge=0)
    thigh_right_cm: float | None = Field(default=None, ge=0)
    neck_cm: float | None = Field(default=None, ge=0)
    shoulders_cm: float | None = Field(default=None, ge=0)
    forearm_cm: float | None = Field(default=None, ge=0)
    calf_cm: float | None = Field(default=None, ge=0)
    notes: str | None = Field(default=None, max_length=500)


class BodyMeasurementResponse(BaseModel):
    """Full measurement entry with database ID."""

    id: int
    date: date
    weight_kg: float | None
    body_fat_pct: float | None
    chest_cm: float | None
    waist_cm: float | None
    hips_cm: float | None
    bicep_left_cm: float | None
    bicep_right_cm: float | None
    thigh_left_cm: float | None
    thigh_right_cm: float | None
    neck_cm: float | None
    shoulders_cm: float | None
    forearm_cm: float | None
    calf_cm: float | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MeasurementProgressPoint(BaseModel):
    """Single data point for measurement progress charts."""

    date: date
    value: float


class MeasurementProgressResponse(BaseModel):
    """Time series data for a specific measurement metric."""

    metric: str
    unit: str
    data: list[MeasurementProgressPoint]
